Internal: Editor V2 - Top bar menu registration infra [ED-9952]

StyleShit opened the pull request.

Nevoss commented on packages/packages/top-bar/src/components/locations/utilities-menu-location.tsx:
```diff
@@ -1,14 +1,16 @@
 export default function UtilitiesMenuLocation() {
-	const injections = useInjectionsOf( LOCATION_UTILITIES_MENU_DEFAULT );
+	const injections = useInjectionsOf(
+		getMenuId( 'utilities', 'default' )
```
```suggestion
		getMenuId( 'utilities' )
```

Make the 'default' default

StyleShit replied:
Done

Nevoss commented on packages/packages/top-bar/src/locations/menus.tsx:
```diff
@@ -1,34 +1,19 @@
 
-type MenuNameMap = Readonly<typeof menuNameMap>;
-export type MenuName = keyof MenuNameMap;
+export type MenuName = keyof Menus;
```
```suggestion
export type MenuName = keyof Menus;
export type MenuGroup<MenuName> = Menus[MenuName];
```

StyleShit replied:
Done

Nevoss commented on packages/packages/top-bar/src/index.ts:
```diff
@@ -1,13 +1,22 @@
 export {
+	getMenuId,
 	registerLink,
 	registerAction,
 	registerToggleAction,
 	injectIntoCanvasDisplay,
 	injectIntoPrimaryAction,
 } from './locations/index';
 
+export type { MenuName, MenuGroup, Menus } from './locations/menus';
+
 export * from './types';
 
+export { default as PopoverMenu } from './components/ui/popover-menu';
+export { default as PopoverMenuItem } from './components/ui/popover-menu-item';
+export { default as ToolbarMenu } from './components/ui/toolbar-menu';
+export { default as ToolbarMenuItem } from './components/ui/toolbar-menu-item';
+export { default as ToolbarMenuToggleItem } from './components/ui/toolbar-menu-toggle-item';
+
```
Let's be sure with @eyalnathan and @rotemee that we want to expose all of those things and make them public, and we will need to support this API.

rami-elementor committed:
Tweak: Accessible navigation arrows in Image Carousel [ED-9947] (#21501)

Co-authored-by: TzviRabinovitch <99259629+TzviRabinovitch@users.noreply.github.com>

elementorbot committed:
Merge branch 'release/3.12.0' into release/3.13.0

elementorbot committed:
Merge branch 'release/3.12.0' into release/3.13.0

yotamselementor committed:
Internal: Stabilize widgets & accordion tests + some alignment (#21559)

## PR Checklist
<!-- 
Please check if your PR fulfills the following requirements:
**Filling out the template is required.** Any pull request that does not
include enough information to be reviewed in a timely manner may be
closed at the maintainers' discretion.
 -->
- [ ] The commit message follows our guidelines:
https://github.com/elementor/elementor/blob/master/.github/CONTRIBUTING.md


## PR Type
What kind of change does this PR introduce?
<!-- Please check the one that applies to this PR using "x" with no
spaces eg: [x]. -->
- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Summary

This PR can be summarized in the following changelog entry:

*

## Description
An explanation of what is done in this PR

*

## Test instructions
This PR can be tested by following these steps:

*

## Quality assurance

- [ ] I have tested this code to the best of my abilities
- [ ] I have added unittests to verify the code works as intended
- [ ] Docs have been added / updated (for bug fixes / features)

Fixes #

elementorbot committed:
Merge branch 'release/3.12.0' into release/3.13.0

elementorbot committed:
Merge branch 'release/3.12.0' into release/3.13.0

elementorbot committed:
Merge branch 'release/3.12.0' into release/3.13.0

StyleShit committed:
Internal: Editor V2 - Top bar menu registration infra [ED-9952]

StyleShit committed:
Fix tests

StyleShit committed:
Merge branch 'release/3.13.0' into internal/ED-9952-top-bar-menus-infra

StyleShit committed:
Revert "Merge branch 'release/3.13.0' into internal/ED-9952-top-bar-menus-infra"

This reverts commit 0b5f3effbb93d5f8fe253b56a94ec5c9dbb0f425.

Nevoss commented on packages/packages/top-bar/src/__tests__/index.test.tsx:
```diff
@@ -254,59 +256,47 @@ describe( '@elementor/top-bar API', () => {
-			},
-		}
-	);
+function makeExampleAction( { onClick = () => null }: { onClick?: () => void } = {} ) {
```
```suggestion
function mockAction( { onClick = () => null }: { onClick?: () => void } = {} ) {
```

StyleShit replied:
Done

Nevoss commented on packages/packages/top-bar/src/__tests__/index.test.tsx:
```diff
@@ -2,15 +2,14 @@ import { useState } from 'react';
 import {
```
Create tests for the `createMenu` API
and separate tests for the components that render the menus if needed

StyleShit replied:
Done

Nevoss commented on packages/packages/top-bar/src/components/locations/main-menu-location.tsx:
```diff
@@ -29,13 +32,13 @@ export default function MainMenuLocation() {
-						.filter( ( injections ) => injections.length )
-						.map( ( injections, index ) => {
+					orderedGroups
+						.filter( ( items ) => items.length )
```
```suggestion
						.filter( ( group ) => group.length )
```

Nevoss commented on packages/packages/top-bar/src/index.ts:
```diff
@@ -1,13 +1,20 @@
-	registerLink,
-	registerAction,
-	registerToggleAction,
+	mainMenu,
```
export createMenu

Nevoss commented on packages/packages/top-bar/src/locations/menus.tsx:
```diff
@@ -0,0 +1,145 @@
+		'default' as const,
+	];
+
+	const registerAction = createRegisterMenuItem( {
+		menuName,
+		menuGroups,
+		component: Action,
+	} );
```
```suggestion
	const [ ... ] = createRegisterMenuItem( {
		menuName,
		menuGroups,
		components: [ Action, ToggleAction, Link ],
	} );
```

Consider

StyleShit committed:
Tests

StyleShit committed:
Merge branch 'release/3.12.0' into internal/ED-9952-top-bar-menus-infra

StyleShit committed:
WIP

StyleShit committed:
Merge branch 'release/3.12.0' into internal/ED-9952-top-bar-menus-infra

Nevoss commented on packages/tests/test-utils/create-mock-menu-actions.tsx:
```diff
@@ -0,0 +1,46 @@
+import { useState } from 'react';
+
+export function createMockMenuAction( { onClick = () => null }: { onClick?: () => void } = {} ) {
```
Add options to overwrite

Nevoss replied:
Add return type, so it will be clear what it retrurns

StyleShit replied:
NO

Nevoss commented on packages/tests/test-utils/create-mock-menu-actions.tsx:
```diff
@@ -0,0 +1,46 @@
+	};
+}
+
+export function createMockMenuToggleAction() {
```
Add options to overwrite

StyleShit replied:
NO

Nevoss commented on packages/tests/test-utils/create-mock-menu-actions.tsx:
```diff
@@ -0,0 +1,46 @@
+	};
+}
+
+export function createMockMenuLink() {
```
Add options to overwrite

StyleShit replied:
NO

Nevoss committed:
Merge branch 'release/3.12.0' into internal/ED-9952-top-bar-menus-infra

StyleShit committed:
Add exports

StyleShit committed:
Merge branch 'release/3.12.0' into internal/ED-9952-top-bar-menus-infra

StyleShit commented on packages/packages/top-bar/src/components/locations/__tests__/menus.test.tsx:
```diff
@@ -0,0 +1,86 @@
+		it( 'should render ordered menu items in a popover', () => {
+			// Arrange.
+			mainMenu.registerAction( {
+				...createMockMenuAction(),
+				name: 'test-action',
```
NO

StyleShit committed:
Merge branch 'release/3.12.0' into internal/ED-9952-top-bar-menus-infra

StyleShit committed:
WIP

StyleShit committed:
Merge branch 'release/3.12.0' into internal/ED-9952-top-bar-menus-infra

StyleShit committed:
Merge branch 'release/3.12.0' into internal/ED-9952-top-bar-menus-infra

StyleShit committed:
WIP

StyleShit committed:
Merge branch 'release/3.12.0' into internal/ED-9952-top-bar-menus-infra
